quality_control_claim/models/qc_inspection.py

```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+# For copyright and license notices, see __openerp__.py file in root directory
+##############################################################################
+from openerp import models, fields, api, exceptions, _
+
+
+class QcInspection(models.Model):
+    _inherit = 'qc.inspection'
+
+    @api.one
+    def _count_claims(self):
+        claim_obj = self.env['crm.claim']
+        cond = [('ref', '=', '%s,%s' % (self._model, self.id))]
+        claims = claim_obj.search(cond)
+        self.claims = len(claims)
+
+    automatic_claims = fields.Boolean(
+        'Automatic Claims', default=False,
+        help="If you want to create one claim when the quality test status is"
+             " 'Quality failed'.")
+    automatic_claims_by_line = fields.Boolean(
+        'Automatic Claims by line', default=False,
+        help="If you want to create one claim per quality test line, when the"
+             " quality test line status is 'Tolerable' or 'Not Tolerable'.")
+    claims = fields.Integer(string="Created claims",
+                            compute='_count_claims', store=False)
+
+    @api.multi
+    def _prepare_inspection_header(self, object_ref, test):
+        result = super(QcInspection, self)._prepare_inspection_header(
+            object_ref, test)
+        result.update({'automatic_claims': test.automatic_claims,
+                       'automatic_claims_by_line':
+                       test.automatic_claims_by_line})
+        return result
+
+    @api.multi
+    def action_approve(self):
+        crm_claim_obj = self.env['crm.claim']
+        category_obj = self.env['crm.case.categ']
+        nc_category_search = [('name', '=', 'NC')]
+        idi_category_search = [('name', '=', 'IDI')]
+        nc_categ = category_obj.search(nc_category_search, limit=1)
+        if not nc_categ:
+            raise exceptions.Warning(
+                _("NC crm case category NOT FOUND"))
+        idi_categ = category_obj.search(idi_category_search, limit=1)
+        if not idi_categ:
+            raise exceptions.Warning(
+                _("IDI crm case category NOT FOUND"))
+        super(QcInspection, self).action_approve()
+        for inspection in self:
+            if inspection.state == 'failed' and inspection.automatic_claims:
+                vals = inspection.init_claim_vals()
+                vals['name'] = _('Quality test %s for product %s'
+                                 ' unsurpassed') % (self.name,
+                                                    self.object_id.name)
+                crm_claim_obj.create(vals)
+            if inspection.automatic_claims_by_line:
+                for line in inspection.inspection_lines:
+                    if line.tolerance_status == 'not_tolerable':
```

Comment:
E independientemente de que se meta lo de Idi y nc en un mod especifico,
por favor no me metais raises de esos. Prefiero que la reclamacion vaya sin
categorizar pero que se cree
si da un raise paras el proceso y probablemente paras la continuidad del
workflow del test. Quitar por favor cualquier raise del codigo que no sea
absoluta y completamente necesario.
El algoritmo de crear los claims deberia ser

Si crear claim por test and test failed -> crear claim
Else
  for each line
       si line no ok -> crear claim

Es decir si el claim se crea por lineas es por cada linea  y si es general
es general. Pero son disjuntos
Y como dice pedro no es por el combo de tolerancia sino por el estado de la
linea

Luego... Ya en el espec de alm... SOLO hay que actualizar la categoria de
los test por linea creados y asignar NC si hiciera falta...
Extension funcional clarisima y dependiente del modulo este + el de
tolerancias.
Mañana lo hablamos si hace falta
El 19/01/2015 20:58, "Pedro M. Baeza" <notifications@github.com> escribió:

> In quality_control_claim/models/qc_inspection.py
> <https://github.com/odoomrp/odoomrp-wip/pull/585#discussion-diff-23182931>
> :
>
> > +                _("NC crm case category NOT FOUND"))
> > +        idi_categ = category_obj.search(idi_category_search, limit=1)
> > +        if not idi_categ:
> > +            raise exceptions.Warning(
> > +                _("IDI crm case category NOT FOUND"))
> > +        super(QcInspection, self).action_approve()
> > +        for inspection in self:
> > +            if inspection.state == 'failed' and inspection.automatic_claims:
> > +                vals = inspection.init_claim_vals()
> > +                vals['name'] = _('Quality test %s for product %s'
> > +                                 ' unsurpassed') % (self.name,
> > +                                                    self.object_id.name)
> > +                crm_claim_obj.create(vals)
> > +            if inspection.automatic_claims_by_line:
> > +                for line in inspection.inspection_lines:
> > +                    if line.tolerance_status == 'not_tolerable':
>
> Esto es del módulo tolerance. Se están mezclando las cosas. Las
> reclamaciones se generan independientemente de la tolerancia. OK: no se
> genera. No OK: se genera si tiene marcado la reclamación automática o se
> pulsa un botón manual de generar reclamación.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/odoomrp/odoomrp-wip/pull/585/files#r23182931>.
>